### PR TITLE
destroy/aws: guard against nils returned by aws

### PR DESCRIPTION
**Pull request by staebler**
Add some additional guards against nil values in objects queried from aws while destroying the cluster.

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1669925

### CODE REVIEW

**Comment by abhinavdahiya** [comment]
https://docs.aws.amazon.com/sdk-for-go/api/aws/#StringValue might be useful.

**Comment by staebler** [comment]
> https://docs.aws.amazon.com/sdk-for-go/api/aws/#StringValue might be useful.

It would be except that I want to log when the load balancer does not have a VPC ID. My strong suspicion is that the load balancer without a VPC ID is one that the installer created--although I haven't rationalized how one could even be created without a VPC ID. I want the log message so that the user is at least informed about the presence of the load balancer.

File: pkg/destroy/aws/aws.go [hunk 1]
```diff
@@ -361,7 +366,7 @@ func getSharedHostedZone(client *route53.Route53, privateID string, logger logru
 			if *zone.Name != privateName {
```
Comment (wking) [inline pkg/destroy/aws/aws.go]: Now that we are allowing input zones of unknown privacy through here as well (vs. requiring them to be private), we probably want a guard against ID matching.

```go
if *zone.Id == privateID {
  continue
}

if *zone.Name != privateName {
```

File: pkg/destroy/aws/aws.go [hunk 2]
```diff
@@ -520,14 +525,19 @@ func deleteEC2InternetGateway(client *ec2.EC2, id string, logger logrus.FieldLog
 
 	for _, gateway := range response.InternetGateways {
 		for _, vpc := range gateway.Attachments {
+			if vpc.VpcId == nil {
+				logger.WithField("gateway", id).Warn("gateway does not have a vpc id")
```
Comment (wking) [inline pkg/destroy/aws/aws.go]: nit: "vpc id" -> "VPC ID"

File: pkg/destroy/aws/aws.go [hunk 3]
```diff
@@ -342,11 +342,16 @@ func getSharedHostedZone(client *route53.Route53, privateID string, logger logru
...
-		return "", errors.Errorf("getShareedHostedZone requires a private ID, but was passed the public %s", privateID)
-	}
 	privateName := *response.HostedZone.Name
 
+	if response.HostedZone.Config != nil && response.HostedZone.Config.PrivateZone != nil {
+		if !*response.HostedZone.Config.PrivateZone {
+			return "", errors.Errorf("getSharedHostedZone requires a private ID, but was passed the public %s", privateID)
+		}
+	} else {
+		logger.Warnf("could not determine whether hosted zone %q is private", privateName)
```
Comment (wking) [inline pkg/destroy/aws/aws.go]: nit: this package likes `WithField`.  What do you think about:

```go
logger.WithField("name", privateName).Warn("could not determine whether hosted zone is private")
```

File: pkg/destroy/aws/aws.go [hunk 4]
```diff
@@ -520,14 +525,19 @@ func deleteEC2InternetGateway(client *ec2.EC2, id string, logger logrus.FieldLog
 
 	for _, gateway := range response.InternetGateways {
 		for _, vpc := range gateway.Attachments {
+			if vpc.VpcId == nil {
+				logger.WithField("gateway", id).Warn("gateway does not have a vpc id")
+				continue
+			}
+			vpcID := *vpc.VpcId
```
Comment (wking) [inline pkg/destroy/aws/aws.go]: You've just checked to ensure `*vpc.VpcId` is non-`nil`.  Why store it in an environment variable?

File: pkg/destroy/aws/aws.go [hunk 5]
```diff
@@ -520,14 +525,19 @@ func deleteEC2InternetGateway(client *ec2.EC2, id string, logger logrus.FieldLog
...
 			_, err := client.DetachInternetGateway(&ec2.DetachInternetGatewayInput{
-				InternetGatewayId: gateway.InternetGatewayId,
-				VpcId:             vpc.VpcId,
+				InternetGatewayId: aws.String(id),
```
Comment (wking) [inline pkg/destroy/aws/aws.go]: `aws.String` is useful for string literals, but when you have a variable, you can just use `&id`.

**Comment by wking** [comment]
I've left a few nits inline, but they should all be minor adjustments.

File: pkg/destroy/aws/aws.go (hunk 3 above)
Comment (staebler) [reply pkg/destroy/aws/aws.go]: "name" does not help much, but I am fine with doing `.WithField("hosted zone", privateName)`

File: pkg/destroy/aws/aws.go (hunk 5 above)
Comment (staebler) [reply pkg/destroy/aws/aws.go]: That assumes that I trust that the callee is not going to modify the value of `id`. But, sure, I can take out these changes.

File: pkg/destroy/aws/aws.go (hunk 3 above)
Comment (wking) [reply pkg/destroy/aws/aws.go]: > "name" does not help much, but I am fine with doing .WithField("hosted zone", privateName)

I think the thing being named will be clear from the "hosted zone" later in the error message.  We'll probably also have the ARN in there too.  For example, a recent CI run has:

```console
$ curl -s https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/1155/pull-ci-openshift-installer-master-e2e-aws/3328/artifacts/e2e-aws/installer/.openshift_install.log | grep hostedzone/
time="2019-02-01T00:26:46Z" level=debug msg="listing AWS hosted zones (page 0)" arn="arn:aws:route53:::hostedzone/Z3BIQ07DC0R5FN" id=Z3BIQ07DC0R5FN
time="2019-02-01T00:26:47Z" level=info msg=Deleted arn="arn:aws:route53:::hostedzone/Z3BIQ07DC0R5FN" id=Z3BIQ07DC0R5FN public zone=/hostedzone/Z328TAU66YDJH7 record set="A ci-op-zbsntbx0-1d3f3-api.origin-ci-int-aws.dev.rhcloud.com."
time="2019-02-01T00:26:47Z" level=info msg=Deleted arn="arn:aws:route53:::hostedzone/Z3BIQ07DC0R5FN" id=Z3BIQ07DC0R5FN record set="A ci-op-zbsntbx0-1d3f3-api.origin-ci-int-aws.dev.rhcloud.com."
time="2019-02-01T00:26:47Z" level=info msg=Deleted arn="arn:aws:route53:::hostedzone/Z3BIQ07DC0R5FN" id=Z3BIQ07DC0R5FN record set="A ci-op-zbsntbx0-1d3f3-etcd-0.origin-ci-int-aws.dev.rhcloud.com."
time="2019-02-01T00:26:47Z" level=info msg=Deleted arn="arn:aws:route53:::hostedzone/Z3BIQ07DC0R5FN" id=Z3BIQ07DC0R5FN record set="A ci-op-zbsntbx0-1d3f3-etcd-1.origin-ci-int-aws.dev.rhcloud.com."
time="2019-02-01T00:26:48Z" level=info msg=Deleted arn="arn:aws:route53:::hostedzone/Z3BIQ07DC0R5FN" id=Z3BIQ07DC0R5FN record set="A ci-op-zbsntbx0-1d3f3-etcd-2.origin-ci-int-aws.dev.rhcloud.com."
time="2019-02-01T00:26:48Z" level=info msg=Deleted arn="arn:aws:route53:::hostedzone/Z3BIQ07DC0R5FN" id=Z3BIQ07DC0R5FN record set="SRV _etcd-server-ssl._tcp.ci-op-zbsntbx0-1d3f3.origin-ci-int-aws.dev.rhcloud.com."
time="2019-02-01T00:26:48Z" level=info msg=Deleted arn="arn:aws:route53:::hostedzone/Z3BIQ07DC0R5FN" id=Z3BIQ07DC0R5FN public zone=/hostedzone/Z328TAU66YDJH7 record set="A \\052.apps.ci-op-zbsntbx0-1d3f3.origin-ci-int-aws.dev.rhcloud.com."
time="2019-02-01T00:26:48Z" level=info msg=Deleted arn="arn:aws:route53:::hostedzone/Z3BIQ07DC0R5FN" id=Z3BIQ07DC0R5FN record set="A \\052.apps.ci-op-zbsntbx0-1d3f3.origin-ci-int-aws.dev.rhcloud.com."
time="2019-02-01T00:26:48Z" level=info msg=Deleted arn="arn:aws:route53:::hostedzone/Z3BIQ07DC0R5FN" id=Z3BIQ07DC0R5FN
```

File: pkg/destroy/aws/aws.go (hunk 5 above)
Comment (wking) [reply pkg/destroy/aws/aws.go]: > That assumes that I trust that the callee is not going to modify the value of `id`.

I'm comfortable with that, and [`aws.String` doesn't make copies anyway][1].

[1]: https://github.com/aws/aws-sdk-go/blob/0943d6551d19b0eeee0b3f342659b5f4e871f025/aws/convert_types.go#L6-L7

File: pkg/destroy/aws/aws.go (hunk 1 above)
Comment (wking) [reply pkg/destroy/aws/aws.go]: > Do you have an opinion about whether we should process input zones of unknown privacy or strictly zones for which we can confirm the privacy?

If the input zone has indeterminate privacy and we find a known-public zone with the same name, I think that's sufficient grounds for assuming the caller gave us a private zone and going ahead to delete entries from the discovered public zone.

File: pkg/destroy/aws/aws.go (hunk 3 above)
Comment (staebler) [reply pkg/destroy/aws/aws.go]: My understanding of the purpose of fields in the logger is to be able to do filtering based upon the fields. A field of name does not help with filtering as much as a field that distinguishes the type of resource being addressed by the log message.

Additionally, a field of name can be ambiguous. For example, take the log message generated at https://github.com/openshift/installer/blob/4907cba3ded1c014f6583f06b51a2e4cf98e8414/pkg/destroy/aws/aws.go#L1008. There are two resources being addressed in the log message, namely the instance profile and the role. Which resource does "name" reference?

File: pkg/destroy/aws/aws.go [hunk 6]
```diff
@@ -520,6 +532,10 @@ func deleteEC2InternetGateway(client *ec2.EC2, id string, logger logrus.FieldLog
 
 	for _, gateway := range response.InternetGateways {
 		for _, vpc := range gateway.Attachments {
+			if vpc.VpcId == nil {
+				logger.WithField("gateway", id).Warn("gateway does not have a VPC ID")
```
Comment (wking) [inline pkg/destroy/aws/aws.go]: The only route to `deleteEC2InternetGateway` already sets the ID in a field [here][1], so we don't need the `WithField` at this point.  You only need `WithField` when you discover new information beyond what the caller gave for the resource you're deleting, or when you switch to look up and remove other resources which are dependencies of the resource you're deleting.

[1]: https://github.com/openshift/installer/blob/5dc2bc9f8b5be67dfef87331ff26f9f82fd09837/pkg/destroy/aws/aws.go#L407

**Comment by staebler** [comment]
@wking CI is green on this. Do you have any more outstanding concerns?

**Comment by abhinavdahiya** [comment]
> /lgtm

/lgtm

**Comment by wking** [comment]
[e2e-aws][1]:

```
Flaky tests:

[Conformance][Area:Networking][Feature:Router] The default ClusterIngress should support default wildcard reencrypt routes through external DNS [Suite:openshift/conformance/parallel/minimal]

Failing tests:

[Feature:Prometheus][Conformance] Prometheus when installed on the cluster should report telemetry if a cloud.openshift.com token is present [Suite:openshift/conformance/parallel/minimal]
[Feature:Prometheus][Conformance] Prometheus when installed on the cluster should start and expose a secured proxy and unsecured metrics [Suite:openshift/conformance/parallel/minimal]
[Feature:Prometheus][Feature:Builds] Prometheus when installed on the cluster should start and expose a secured proxy and verify build metrics [Suite:openshift/conformance/parallel]
```

/retest

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/1151/pull-ci-openshift-installer-master-e2e-aws/3497/build-log.txt

**Comment by wking** [comment]
[e2e-aws][1]:

```
Flaky tests:

[Conformance][Area:Networking][Feature:Router] The default ClusterIngress should support default wildcard reencrypt routes through external DNS [Suite:openshift/conformance/parallel/minimal]
[sig-storage] In-tree Volumes [Driver: aws] [Testpattern: Dynamic PV (default fs)] subPath should support file as subpath [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] In-tree Volumes [Driver: aws] [Testpattern: Dynamic PV (default fs)] subPath should support readOnly directory specified in the volumeMount [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] In-tree Volumes [Driver: aws] [Testpattern: Dynamic PV (default fs)] volumes should be mountable [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] In-tree Volumes [Driver: aws] [Testpattern: Dynamic PV (ext3)] volumes should be mountable [Suite:openshift/conformance/parallel] [Suite:k8s]

Failing tests:

[sig-apps] StatefulSet [k8s.io] Basic StatefulSet functionality [StatefulSetBasic] should not deadlock when a pod's predecessor fails [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-apps] StatefulSet [k8s.io] Basic StatefulSet functionality [StatefulSetBasic] should perform rolling updates and roll backs of template modifications with PVCs [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-apps] StatefulSet [k8s.io] Basic StatefulSet functionality [StatefulSetBasic] should provide basic identity [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-scheduling] Multi-AZ Clusters should spread the pods of a replication controller across zones [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-scheduling] Multi-AZ Clusters should spread the pods of a service across zones [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] Dynamic Provisioning DynamicProvisioner should provision storage with different parameters [Suite:openshift/conformance/parallel] [Suite:k8s]
```

Flakes moving in the wrong direction ;).

/retest

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/1151/pull-ci-openshift-installer-master-e2e-aws/3499/build-log.txt

**Comment by wking** [comment]
[e2e-aws][1]:

```
Failing tests:

[Conformance][Area:Networking][Feature:Router] The HAProxy router should override the route host for overridden domains with a custom value [Suite:openshift/conformance/parallel/minimal]
[Conformance][Area:Networking][Feature:Router] The HAProxy router should override the route host with a custom value [Suite:openshift/conformance/parallel/minimal]
[Conformance][Area:Networking][Feature:Router] The HAProxy router should serve a route that points to two services and respect weights [Suite:openshift/conformance/parallel/minimal]
[Conformance][Area:Networking][Feature:Router] The HAProxy router should serve routes that were created from an ingress [Suite:openshift/conformance/parallel/minimal]
```

/retest

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/1151/pull-ci-openshift-installer-master-e2e-aws/3506/build-log.txt

**Comment by wking** [comment]
[e2e-aws][1]:

```'
Flaky tests:

[sig-storage] PersistentVolumes NFS with Single PV - PVC pairs create a PVC and non-pre-bound PV: test write access [Suite:openshift/conformance/parallel] [Suite:k8s]

Failing tests:

[Conformance][Area:Networking][Feature:Router] The default ClusterIngress should support default wildcard reencrypt routes through external DNS [Suite:openshift/conformance/parallel/minimal]
```

/retest

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/1151/pull-ci-openshift-installer-master-e2e-aws/3510/build-log.txt

**Comment by wking** [comment]
[e2e-aws hung][1].

/retest

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/1151/pull-ci-openshift-installer-master-e2e-aws/3513/build-log.txt

**Comment by wking** [comment]
[e2e-aws][1]:

```
level=error msg="\t* module.vpc.aws_route.to_nat_gw[5]: 1 error occurred:"
level=error msg="\t* aws_route.to_nat_gw.5: Error creating route: timeout while waiting for state to become 'success' (timeout: 2m0s)"
```

/retest

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/1151/pull-ci-openshift-installer-master-e2e-aws/3526/build-log.txt

**Pull request merged**